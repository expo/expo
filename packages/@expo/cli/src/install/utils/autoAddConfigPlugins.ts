import { ExpoConfig } from '@expo/config';
import {
  normalizeStaticPlugin,
  resolveConfigPluginFunctionWithInfo,
} from '@expo/config-plugins/build/utils/plugin-resolver';
import { getAutoPlugins } from '@expo/prebuild-config';

import { attemptAddingPluginsAsync } from '../../utils/modifyConfigPlugins';

const debug = require('debug')('expo:install:config-plugins') as typeof console.log;

const AUTO_PLUGINS = getAutoPlugins();

/**
 * Resolve if a package has a config plugin.
 * For sanity, we'll only support config plugins that use the `app.config.js` entry file,
 * this is because a package like `lodash` could be a "valid" config plugin and break the prebuild process.
 *
 * @param projectRoot
 * @param packageName
 * @returns
 */
function packageHasConfigPlugin(projectRoot: string, packageName: string) {
  try {
    const info = resolveConfigPluginFunctionWithInfo(projectRoot, packageName);
    if (info.isPluginFile) {
      return info.plugin;
    }
  } catch {}
  return false;
}

/**
 * Get a list of plugins that were are supplied as string module IDs.
 * @example
 * ```json
 * {
 *   "plugins": [
 *     "expo-camera",
 *     ["react-native-firebase", ...]
 *   ]
 * }
 * ```
 *   ↓ ↓ ↓ ↓ ↓ ↓
 *
 * `['expo-camera', 'react-native-firebase']`
 *
 */
export function getNamedPlugins(plugins: NonNullable<ExpoConfig['plugins']>): string[] {
  const namedPlugins: string[] = [];
  for (const plugin of plugins) {
    try {
      // @ts-ignore
      const [normal] = normalizeStaticPlugin(plugin);
      if (typeof normal === 'string') {
        namedPlugins.push(normal);
      }
    } catch {
      // ignore assertions
    }
  }
  return namedPlugins;
}

/** Attempts to ensure that non-auto plugins are added to the `app.json` `plugins` array when modules with Expo Config Plugins are installed. */
export async function autoAddConfigPluginsAsync(
  projectRoot: string,
  exp: Pick<ExpoConfig, 'plugins'>,
  packages: string[]
) {
  debug('Checking config plugins...');

  const currentPlugins = exp.plugins || [];
  const normalized = getNamedPlugins(currentPlugins);

  debug(`Existing plugins: ${normalized.join(', ')}`);

  const plugins = packages.filter((pkg) => {
    if (normalized.includes(pkg)) {
      // already included in plugins array
      return false;
    }
    // Check if the package has a valid plugin. Must be a well-made plugin for it to work with this.
    const plugin = packageHasConfigPlugin(projectRoot, pkg);

    debug(`Package "${pkg}" has plugin: ${!!plugin}` + (plugin ? ` (args: ${plugin.length})` : ''));

    if (AUTO_PLUGINS.includes(pkg)) {
      debug(`Package "${pkg}" is an auto plugin, skipping...`);
      return false;
    }

    return !!plugin;
  });

  await attemptAddingPluginsAsync(projectRoot, exp, plugins);
}
