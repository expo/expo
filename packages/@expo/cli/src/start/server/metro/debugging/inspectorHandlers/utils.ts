import type WS from 'ws';

import { DebuggerResponse, DeviceResponse } from './types';

/**
 * Helper function to respond to a message from the debugger or device.
 * The return value is used to stop the message propagation, "canceling" further handling.
 *
 * @example ```
 *  return respond<DeviceResponse<CDP>>(socket, { id: message.id, result: {} });
 * ```
 */
export function respond<T = DeviceResponse | DebuggerResponse>(socket: WS, message: T) {
  socket.send(JSON.stringify(message));
  return true;
}

/** Known compatible debuggers that require specific workarounds */
export type DebuggerType = 'chrome' | 'vscode' | 'unknown';

// Patterns to test against user agents
const CHROME_USER_AGENT = /chrome/i;
const VSCODE_USER_AGENT = /vscode/i;

/**
 * Determine the debugger type based on the known user agent.
 */
export function getDebuggerType(userAgent?: string | null): DebuggerType {
  if (userAgent && CHROME_USER_AGENT.test(userAgent)) return 'chrome';
  if (userAgent && VSCODE_USER_AGENT.test(userAgent)) return 'vscode';
  return 'unknown';
}
