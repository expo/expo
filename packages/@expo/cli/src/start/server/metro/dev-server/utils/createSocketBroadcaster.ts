import type { RawData as WebSocketRawData } from 'ws';

import type { SocketId, SocketMap } from './createSocketMap';

const debug = require('debug')('expo:metro:dev-server:broadcaster') as typeof console.log;

export function createBroadcaster(sockets: SocketMap) {
  return function broadcast(senderSocketId: SocketId | null, message: string | WebSocketRawData) {
    // Ignore if there are no connected sockets
    if (!sockets.size) return;

    for (const [socketId, socket] of sockets) {
      if (socketId === senderSocketId) continue;

      try {
        socket.send(message);
      } catch (error) {
        debug(`Failed to broadcast message to socket "${socketId}"`, error);
      }
    }
  };
}
