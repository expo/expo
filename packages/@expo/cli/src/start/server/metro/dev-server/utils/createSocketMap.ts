import type { WebSocket } from 'ws';

const debug = require('debug')('expo:metro:dev-server:socketmap') as typeof console.log;

export type SocketId = string;
export type SocketMap = Map<string, WebSocket>;

export function createSocketMap() {
  const map: SocketMap = new Map();
  const createId = createSocketIdFactory();

  const registerSocket = (socket: WebSocket) => {
    const id = createId();
    map.set(id, socket);
    return {
      id,
      terminate: () => {
        map.delete(id);
        socket.removeAllListeners();
        socket.terminate();
      },
    };
  };

  const findSocket = (id: SocketId): WebSocket | null => {
    const socket = map.get(id);
    if (!socket) debug(`No connected socket found with ID: ${id}`);
    return socket ?? null;
  };

  return { map, registerSocket, findSocket };
}

function createSocketIdFactory() {
  let nextId = 0;
  return () => `socket#${nextId++}`;
}
