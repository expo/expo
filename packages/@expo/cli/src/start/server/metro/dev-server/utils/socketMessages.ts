import type { RawData as WebSocketRawData } from 'ws';

const debug = require('debug')('expo:metro:dev-server:messages') as typeof console.log;

/** The current websocket-based communication between Metro, CLI, and client devices */
const PROTOCOL_VERSION = 2;

/**
 * Parse the incoming raw message data and return the parsed object.
 * This returns null if the protocol version did not match expected version.
 */
export function parseRawMessage<T = Record<string, any>>(
  data: WebSocketRawData,
  isBinary: boolean
): null | T {
  if (isBinary) return null;

  try {
    const { version, ...message } = JSON.parse(data.toString()) ?? {};
    if (version === PROTOCOL_VERSION) {
      return message;
    }

    debug(
      `Received message protocol version did not match supported "${PROTOCOL_VERSION}", received: ${message.version}`
    );
  } catch (error) {
    debug(`Failed to parse message: ${error}`);
  }

  return null;
}

/**
 * Serialize any of the messages to send over websockets.
 * This adds the protocol version to the message.
 */
export function serializeMessage(message: Record<string, any>): string {
  return JSON.stringify({ ...message, version: PROTOCOL_VERSION });
}
