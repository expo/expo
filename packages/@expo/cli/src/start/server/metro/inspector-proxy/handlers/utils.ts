import type WS from 'ws';

import { DebuggerResponse, DeviceResponse } from './types';

/**
 * Helper function to respond to a message from the debugger or device.
 * The return value is used to stop the message propagation, "canceling" further handling.
 *
 * @example ```
 *  return respond<DeviceResponse<CDP>>(socket, { id: message.id, result: {} });
 * ```
 */
export function respond<T = DeviceResponse | DebuggerResponse>(socket: WS, message: T) {
  socket.send(JSON.stringify(message));
  return true;
}
