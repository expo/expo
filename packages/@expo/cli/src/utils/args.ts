// Common utilities for interacting with `args` library.
// These functions should be used by every command.
import arg from 'arg';
import chalk from 'chalk';
import { existsSync } from 'fs';
import { resolve } from 'path';

import * as Log from '../log';

/**
 * Parse the first argument as a project directory.
 *
 * @returns valid project directory.
 */
export function getProjectRoot(args: arg.Result<arg.Spec>) {
  const projectRoot = resolve(args._[0] || '.');

  if (!existsSync(projectRoot)) {
    Log.exit(`Invalid project root: ${projectRoot}`);
  }

  return projectRoot;
}

/**
 * Parse args and assert unknown options.
 *
 * @param schema the `args` schema for parsing the command line arguments.
 * @param argv extra strings
 * @returns processed args object.
 */
export function assertArgs(schema: arg.Spec, argv?: string[]): arg.Result<arg.Spec> {
  return assertWithOptionsArgs(schema, { argv });
}

export function assertWithOptionsArgs(
  schema: arg.Spec,
  options: arg.Options
): arg.Result<arg.Spec> {
  try {
    return arg(schema, options);
  } catch (error: any) {
    // Ensure unknown options are handled the same way.
    if (error.code === 'ARG_UNKNOWN_OPTION') {
      Log.exit(error.message, 1);
    }
    // Otherwise rethrow the error.
    throw error;
  }
}

export function printHelp(info: string, usage: string, options: string, extra: string = ''): never {
  Log.exit(
    chalk`
  {bold Info}
    ${info}

  {bold Usage}
    {dim $} ${usage}

  {bold Options}
    ${options.split('\n').join('\n    ')}
` + extra,
    0
  );
}
