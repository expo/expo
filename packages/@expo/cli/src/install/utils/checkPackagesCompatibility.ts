// note(Simek): https://github.com/react-native-community/directory/blob/main/pages/api/libraries/check.ts
import chalk from 'chalk';

import { Log } from '../../log';
import { fetch } from '../../utils/fetch';
import { learnMore } from '../../utils/link';

export type ReactNativeDirectoryCheckResult = {
  unmaintained: boolean;
  newArchitecture: 'supported' | 'unsupported' | 'untested';
};

const ERROR_MESSAGE =
  'Unable to fetch compatibility data from React Native Directory. Skipping check.';

export async function checkPackagesCompatibility(packages: string[]) {
  try {
    const response = await fetch('https://reactnative.directory/api/libraries/check', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({ packages }),
    });

    if (!response.ok) {
      Log.log(chalk.gray(ERROR_MESSAGE));
    }

    const packageMetadata = (await response.json()) as Record<
      string,
      ReactNativeDirectoryCheckResult
    >;

    const incompatiblePackages = packages.filter(
      (packageName) => packageMetadata[packageName]?.newArchitecture === 'unsupported'
    );

    if (incompatiblePackages.length) {
      Log.warn(
        chalk.yellow(
          `${chalk.bold('Warning')}: ${formatPackageNames(incompatiblePackages)} do${incompatiblePackages.length > 1 ? '' : 'es'} not support the New Architecture. ${learnMore('https://docs.expo.dev/guides/new-architecture/')}`
        )
      );
    }
  } catch {
    Log.log(chalk.gray(ERROR_MESSAGE));
  }
}

function formatPackageNames(incompatiblePackages: string[]) {
  if (incompatiblePackages.length === 1) {
    return incompatiblePackages.join();
  }

  const lastPackage = incompatiblePackages.pop();
  return `${incompatiblePackages.join(', ')} and ${lastPackage}`;
}
