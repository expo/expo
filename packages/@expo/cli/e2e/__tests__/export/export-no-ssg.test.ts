import execa from 'execa';
import klawSync from 'klaw-sync';
import path from 'path';

import { runExportSideEffects } from './export-side-effects';
import { bin, getRouterE2ERoot } from '../utils';
import JsonFile from '@expo/json-file';

runExportSideEffects();

describe('export-no-ssg', () => {
  const projectRoot = getRouterE2ERoot();
  const outputDir = path.join(projectRoot, 'dist-server-no-ssg');

  beforeAll(
    async () => {
      console.time('export-server');
      await execa(
        'node',
        [bin, 'export', '-p', 'web', '-p', 'ios', '--output-dir', 'dist-server-no-ssg', '--no-ssg'],
        {
          cwd: projectRoot,
          env: {
            NODE_ENV: 'production',
            EXPO_USE_STATIC: 'server',
            E2E_ROUTER_SRC: 'server',
            E2E_ROUTER_ASYNC: 'development',
            EXPO_USE_FAST_RESOLVER: 'true',
          },
        }
      );
      console.timeEnd('export-server');
    },
    // Could take 45s depending on how fast the bundler resolves
    560 * 1000
  );

  function getFiles() {
    // List output files with sizes for snapshotting.
    // This is to make sure that any changes to the output are intentional.
    // Posix path formatting is used to make paths the same across OSes.
    return klawSync(outputDir)
      .map((entry) => {
        if (entry.path.includes('node_modules') || !entry.stats.isFile()) {
          return null;
        }
        return path.posix.relative(outputDir, entry.path);
      })
      .filter(Boolean);
  }

  it(
    'has expected files',
    async () => {
      // Request HTML

      // List output files with sizes for snapshotting.
      // This is to make sure that any changes to the output are intentional.
      // Posix path formatting is used to make paths the same across OSes.
      const files = getFiles();

      // The wrapper should not be included as a route.
      expect(files).not.toContain('server/+html.html');
      expect(files).not.toContain('server/_layout.html');

      // Has routes.json
      expect(files).toContain('server/_expo/routes.json');

      // Has functions
      expect(files).toContain('server/_expo/functions/methods+api.js');
      expect(files).toContain('server/_expo/functions/methods+api.js.map');
      expect(files).toContain('server/_expo/functions/api/[dynamic]+api.js');
      expect(files).toContain('server/_expo/functions/api/[dynamic]+api.js.map');
      expect(files).toContain('server/_expo/functions/api/externals+api.js');
      expect(files).toContain('server/_expo/functions/api/externals+api.js.map');

      // TODO: We shouldn't export this
      expect(files).toContain('server/_expo/functions/api/empty+api.js');
      expect(files).toContain('server/_expo/functions/api/empty+api.js.map');

      // Has single variation of group file
      expect(files).not.toContain('server/(alpha)/index.html');
      expect(files).not.toContain('server/(alpha)/beta.html');
      expect(files).not.toContain('server/beta.html');

      // Injected by framework
      expect(files).not.toContain('server/_sitemap.html');
      expect(files).not.toContain('server/+not-found.html');

      // Normal routes
      expect(files).toContain('client/index.html');
      expect(files).not.toContain('server/blog/[post].html');

      const json = await JsonFile.readAsync(path.join(outputDir, 'server/_expo/routes.json'));
      expect((json.apiRoutes as any[]).length).toBeGreaterThan(0);
      expect(json.htmlRoutes).toEqual([]);
      expect(json.notFoundRoutes).toEqual([]);
    },
    5 * 1000
  );
});
