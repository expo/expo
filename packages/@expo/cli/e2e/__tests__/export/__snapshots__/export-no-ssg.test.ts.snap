// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`export-no-ssg has expected routes manifest entries 1`] = `
{
  "apiRoutes": [
    {
      "file": "_expo/functions/(a,b)/multi-group-api+api.js",
      "namedRegex": "^(?:/\\((?:a|b)\\))?/multi\\-group\\-api(?:/)?$",
      "page": "/(b)/multi-group-api",
      "routeKeys": {},
    },
    {
      "file": "_expo/functions/methods+api.js",
      "namedRegex": "^/methods(?:/)?$",
      "page": "/methods",
      "routeKeys": {},
    },
    {
      "file": "_expo/functions/api/json+api.js",
      "namedRegex": "^/api/json(?:/)?$",
      "page": "/api/json",
      "routeKeys": {},
    },
    {
      "file": "_expo/functions/api/empty+api.js",
      "namedRegex": "^/api/empty(?:/)?$",
      "page": "/api/empty",
      "routeKeys": {},
    },
    {
      "file": "_expo/functions/api/error+api.js",
      "namedRegex": "^/api/error(?:/)?$",
      "page": "/api/error",
      "routeKeys": {},
    },
    {
      "file": "_expo/functions/api/env-vars+api.js",
      "namedRegex": "^/api/env\\-vars(?:/)?$",
      "page": "/api/env-vars",
      "routeKeys": {},
    },
    {
      "file": "_expo/functions/api/redirect+api.js",
      "namedRegex": "^/api/redirect(?:/)?$",
      "page": "/api/redirect",
      "routeKeys": {},
    },
    {
      "file": "_expo/functions/api/externals+api.js",
      "namedRegex": "^/api/externals(?:/)?$",
      "page": "/api/externals",
      "routeKeys": {},
    },
    {
      "file": "_expo/functions/api/problematic+api.js",
      "namedRegex": "^/api/problematic(?:/)?$",
      "page": "/api/problematic",
      "routeKeys": {},
    },
    {
      "file": "_expo/functions/matching-route/alpha+api.js",
      "namedRegex": "^/matching\\-route/alpha(?:/)?$",
      "page": "/matching-route/alpha",
      "routeKeys": {},
    },
    {
      "file": "_expo/functions/api/[dynamic]+api.js",
      "namedRegex": "^/api/(?<dynamic>[^/]+?)(?:/)?$",
      "page": "/api/[dynamic]",
      "routeKeys": {
        "dynamic": "dynamic",
      },
    },
    {
      "file": "_expo/functions/api/a/[...spread]+api.js",
      "namedRegex": "^/api/a(?:/(?<spread>.+?))?(?:/)?$",
      "page": "/api/a/[...spread]",
      "routeKeys": {
        "spread": "spread",
      },
    },
  ],
  "htmlRoutes": [],
  "notFoundRoutes": [],
}
`;
